### PR TITLE
Add number of allocations to time/timed macros, fix realloc calculations

### PR DESCRIPTION
**Pull request by ScottPJones**
This adds counting of the number of allocations to `@time` and `@timed` (`@timed` also now returns the number of `free`s, and `realloc`s).
It also fixes a bug that happens if you use realloc to shrink the size of  something, it would subtract from allocd_bytes, instead of adding to freed_bytes.

### CODE REVIEW

**Comment by StefanKarpinski** [comment]
This seems like a reasonable way to do this for now, but it might be good in the future to refactor things so that there are fewer tiny functions to get info about allocations. Maybe a struct that can be returned.

**Comment by carnaval** [comment]
You are not counting pool allocations, which are 99% of the workload in most cases. It is not strictly a malloc call but I think people would expect it to be taken into account.

I'm not sure that the number of free calls is very informative for the same reason. We could expose the number of freed objects instead but it requires a bit of addtional bookeeping in the pool sweeping. I mostly (perhaps cowardly) don't want people trying to second guess the gc heuristics and argue to change them. The number of allocation only depends on your code whereas the number of freed objects depends a lot on the gc's decisions.

**Comment by carnaval** [comment]
@StefanKarpinski agreed

**Comment by ScottPJones** [comment]
@StefanKarpinski Very good point, I was just following what was there before... refactoring would be nice... I wanted to refactor timed / time, but I'm not sure how (yet) with macros... more investigation!

**Comment by ScottPJones** [comment]
@carnaval I was just changing the places where allocd_bytes was being updated... I missed this pooled allocation? If you can point me where to fix it, I will, thanks!

**Comment by carnaval** [comment]
https://github.com/JuliaLang/julia/blob/master/src/gc.c#L1030

**Comment by carnaval** [comment]
I think the number of allocation could use a little pretty printing since it is likely to get very large, as you will probably experience once you add the pool stuff.

**Comment by ScottPJones** [comment]
@carnaval I ended up splitting up the display of malloc/realloc/pool, they have very different performance characteristics (realloc may mean copying, etc.).  Is there a nice function already to get "K","M","G" to display counts prettily?  Thanks!

**Comment by StefanKarpinski** [comment]
I really think this is getting out of hand. It would be fine if we provide a way or getting this info for those who really need to know, but this is *way* too much info to display by default.

**Comment by ScottPJones** [comment]
@StefanKarpinski Adding the number of allocations was from @vtjnash 's suggestion on julia-users.
I found a lot of value in seeing the information about allocations, it was actually more useful than the raw amount allocated for what I was doing, however, since I've added the information to the `@timed` macro, I assume someone could easily print out just the pieces they are interested in using that.  I can remove the extra display from `@time`, if that's what everybody (or the majority) want.

**Comment by StefanKarpinski** [comment]
I'm fine with the number of allocations, but separating out the number of reallocs is too much. And what is this "pool" value? Displaying it with slashes between means there's zero chance that someone can understand what's going on just from reading the output of `@time`.

**Comment by ScottPJones** [comment]
So, would you have mallocs+reallocs, or mallocs+reallocs+pool, displayed as a single number in `@time`?  The pool allocations are ones that are handled without having to resort to a C malloc call.
reallocs of course are interesting to see if you have code that is allocating an array and growing it, instead of allocating it once, part of what I changed to speed things up for UTF conversions... but I'm happy just getting them separately from `@timed`.

**Comment by StefanKarpinski** [comment]
I don't know, but this is too much information. We should either display a multiline detailed output here that's clear and readable or introduce another mechanism for getting this information. I suspect that returning a struct from `@timed` would be a good idea – this can be the same shape of struct that the C code uses to collect the data. It can even be made iterable to avoid breaking old code.

**Comment by ScottPJones** [comment]
`@timed` just returns a tuple, I simply added extra values at the end... would returning the information as a structure definitely be better? (I'm happy to do it however, just want the functionality available...)

**Comment by StefanKarpinski** [comment]
I made that comment at the point where you were also printing all the information. I do think that returning a structure might be more reasonable at this point. Returning that many values from a function is getting unwieldy.

**Comment by carnaval** [comment]
I agree with Stefan. I think a summary of total_number_of_alloc_events and total_size_allocated is sufficient. I don't think we have a facility for big number formatting in Base.

I'm not convinced by the realloc argument. I'm not against providing it as a separate function but I think just having the summary is what people need. Remember that a lot of people are using `@time` for actual timing and the GC details are not of interest to them (well except when it's the bottleneck I guess).

**Comment by carnaval** [comment]
Minor style nitpick but could you rename those counters to `num_` instead of `cnt_` (just for consistency with the rest, even if admittedly there are many inconsistent naming in there). I also think that `num_pool_alloc` is clearer than `num_pool`. Thanks !

**Comment by ScottPJones** [comment]
All good stuff everyone!  I'm trying to get this done ASAP so I can get back to some "real" work :grinning:
@StefanKarpinski I'll put it in a structure, no problem, for coming back from C, do you want @timed to also return the structure?  Currently it returns a tuple, I don't want to break anybody who uses it... should I just add the structure as the last value in the tuple?
@carnaval Yes, I'd meant to do that... but, if you look at the number of issues I've made (and am trying to fix myself), I slipped up!  Will do (cnt -> num).  reallocs are only returned in `@timed` now, not displayed in `@time`.

**Comment by ScottPJones** [comment]
Help!  I must have messed things up with Git!  This is only supposed to be 3 files! I updated it to remove a trailing space on one line... and now... :cry:

**Comment by pao** [comment]
Oooh. Got a merge. Nasty. The following is not recommended, but is also exactly what I'd do to fix.

```
git branch pao-broke-my-repository # always good to make a backup of your current situation
git reset --hard b7f9286 # note any changes not checked in will be lost
git cherry-pick a3d8a1c # but we do want that whitespace fixed
```

**Comment by ScottPJones** [comment]
@pao You are a real lifesaver!  That looks like it did the trick, thanks!

**Comment by ScottPJones** [comment]
Any reason this cannot be merged in?  I believe I've addressed all of the requests for changes.
Thanks, Scott

**Comment by tkelman** [comment]
Anywhere that we have a struct in C and a type in Julia that need to be kept exactly in sync, it would be good to mark both with a comment pointing to the other - "If you change this, be sure to make the corresponding change in ____"

**Comment by ScottPJones** [comment]
OK, very good point!  Will add that now... anything else?

**Comment by tkelman** [comment]
Whoops, didn't notice this at first, but speaking of keeping them in sync, is it just me or are the last 3 elements in different orders?

**Comment by ScottPJones** [comment]
@tkelman Thanks very much for catching that! I need more sleep or more caffeine before I commit stuff!

**Comment by carnaval** [comment]
Last few points, then I think you can squash and this is good to go :
- could you make the struct immutable ? It will require making a new one in the diff function. It may be stupid but I don't like the idea of allocating gc memory in a place where we are measuring it. Making it immutable will make this not hit the gc at all.
- I think the wording can be made more compact like "bla allocations (blu MB)" instead of repeating allocated
- it will be quickly unreadable if you don't pretty print the number of allocation events, this can go to billions easily

**Comment by ScottPJones** [comment]
OK, all good points.  About making the struct immutable, doesn't that mean you'd always have to do an allocation, instead of just allocating 2 structs at the beginning of a loop (if you are reporting the differences in a loop)?  I'm still learning all the ins and outs of optimizing Julia for mem allocations...
I also had wondered about getting rid of the repeated "allocations" "allocated", your idea looks good. Thanks!

**Comment by carnaval** [comment]
If a type is immutable _and_ known to the compiler without ambiguities (which is very likely to be the case here) it is passed around by copies and never uses dynamic memory allocation. The only case where we do allocate memory for such objects is when it must be passed to code which may not be aware of the type in advance and require boxing it.

**Comment by ScottPJones** [comment]
Cool, thanks! I'll change that as soon as I back home

**Comment by ScottPJones** [comment]
OK, I'm mostly done, but what pretty printing do you think is best? (the current one isn't that good [sorry, whoever wrote it, my comments are never meant to be personal!, it truncates instead of rounding, for example, and says "kB" and "MB"... not consistent with common computer usage).
For allocations, by 1000s: "", "k", "M", "G", "T", "P" (i.e. SI format),
and memory, by 1024s: "B", "KB", "MB", "GB", "TB", "PB".

**Comment by ScottPJones** [comment]
How does this look everybody?
```julia
julia> @Base.time for i=1:1000000 ; x = utf16("this is a test\u10ffff") ; end
379msec, 6000k allocations (250000KB)
(6.26% gc time in 11 pauses with 0 full sweep)
```

**Comment by ScottPJones** [comment]
@carnaval Is this good enough to be merged now?  Thanks!

**Comment by ScottPJones** [comment]
Here is the output on my "integration" branch, which has my conversion changes:
```julia
julia> @time for i=1:1000000 ; x = utf16("this is a test\u10ffff") ; end
101msec, 2000k allocations (125000KB)
(4.82% gc time in 6 pauses with 0 full sweep)

julia> @timev for i=1:1000000 ; x = utf16("this is a test\u10ffff") ; end
113msec, 2000k allocations (125000KB)
(4.15% gc time in 5 pauses with 0 full sweep)
allocated:       13311957
freed:           0
mallocs:         0
reallocs:        0
poolallocs:      2000000
free calls:      0
total allocated: 114688043
since sweep:     0
```
This shows that without my other changes, it is doing 3x the number of allocations for a 16 character string... (and people wondered why I thought it was `O(n^2)` :wink: )

**Comment by ScottPJones** [comment]
I may have to raise another issue for this... but why isn't there a "pool realloc"?  This shows that these always have to copy... whereas if they had been malloc'ed then realloc'ed, the copying might have been avoided... [there may be a simple answer to this, I haven't had time to go through all of the memory allocation scheme yet... sorry for my lack of research before posting! :wink: ]

**Comment by kmsquire** [comment]
In Julia v0.3, this prints
```julia
julia> @time for i=1:1000000 ; x = utf16("this is a test\u10ffff") ; end
elapsed time: 1.295873191 seconds (216000000 bytes allocated, 24.57% gc time)
```

Except that bytes is easier to parse in KB/MB/GB, I think that this prints the right amount of information for most people.  

Adding `in 11 pauses with 0 full sweep` was probably useful to some developers for understanding and debugging the new gc.  It was a little distracting (I'm not familiar with `pauses` and `full sweep`), but I have a vague notion of what it means and could ignore it.

Additionally adding `2000k allocations` right after the time and splitting it into two lines moves it to "too much stuff I need to parse and/or ignore".

I can see how all of this information could be useful, but could it be moved to a different macro, say `@allocated`, leaving `@time` sweet and simple?

**Comment by ScottPJones** [comment]
I wasn't aware of how it worked in 0.3 (having pretty much only used 0.4 since starting 6 weeks ago).
I can remove the second line (and move the % gc time) to the 1st line.
I think "elapsed time: " is redundant, since you are calling a `@time` macro...
I do think having the information on allocations is very important for anybody trying to figure out performance though...  I also tried to make the # nanoseconds/seconds easier to parse / compare [I hate the way it currently uses scientific notation for the time... very hard to compare]
The verbose form, `@timev` already has all of the extra information... very useful IMO for performance debugging...

**Comment by kmsquire** [comment]
> I think "elapsed time: " is redundant, since you are calling a `@time` macro...

To some extent, it is redundant, but it also helps set off the time from other output that might be generated within the timed code.

**Comment by kmsquire** [comment]
One other suggestion: how about something like:

```julia
101 ms (23 Mb allocated from 2.3 million allocations, in 11 pauses with 0 full sweep)
```

(I'm having a *deja vu*--I thought something like this was previously suggested, but I don't see it above.)

**Comment by ScottPJones** [comment]
I find the # of pauses / sweeps not really necessary all the time (the % gctime I think is very useful).
I'd already added a space between the number and the "ms" (testing it with some other minor cleanups now)

**Comment by kmsquire** [comment]
Agree with `%gc time` being useful--sorry I didn't put that above.  (Now edited inline.)

While I agree that the number of allocations can be useful, I find `2000k allocations` in your previous example hard to parse.  Is it "two-thousand thousand" allocations?  I also feel that it would work better inside the parentheses--this is a `@time` macro, not an `@allocated` macro.  Which already does exist, and could probably use similar updates:

```julia
julia> @allocated for i=1:1000000 ; x = utf16("this is a test\u10ffff") ; end
256000000
```

**Comment by carnaval** [comment]
totally fine with removing the number of partial/full sweeps. I added this sneakily in the big gc commit (0bfe05df) so probably didn't get much review. I'm notoriously bad at UI so I'll let the final word to someone else. IMO the n# of allocations should be the main info since it is usually more correlated to perf and the size is apparently confusing people (where did my 12G of RAM go ?).

**Comment by ScottPJones** [comment]
About the "2000k allocations", that's just because that happened to be a rather contrived example, with a power of 10 number of iterations...
Here is what I have currently:
```julia
julia> @timev for i=1:1234 ; x = utf16("this is a test\u10ffff") ; end
414 µsec (7404 allocations: 315904B)
elapsed time:    413504 nanoseconds
bytes allocated: 315904
allocated:       315904
poolallocs:      7404

julia> @time for i=1:1234 ; x = utf16("this is a test\u10ffff") ; end
418 µsec (7404 allocations: 315904B)
```

**Comment by kmsquire** [comment]
+1!

**Comment by ScottPJones** [comment]
Here it is again, with a few more improvements:
```julia
julia> @timev for i=1:2 ; x = utf16("this is a test\u10ffff") ; end
5537 nsec (12 allocations: 512 bytes)
elapsed time:    5537 nanoseconds
bytes allocated: 512
allocated:       512
poolallocs:      12

julia> @time for i=1:2 ; x = utf16("this is a test\u10ffff") ; end
4683 nsec (12 allocations: 512 bytes)

julia> @time for i=1:1000 ; x = length("this is a test\u10ffff") ; end
11318 nsec

julia> @time for i=1:1000000 ; x = length("this is a test\u10ffff") ; end
11908 µsec

julia> @time for i=1:1000000000 ; x = length("this is a test\u10ffff") ; end
11931 msec

julia> @time for i=1:10000000000 ; x = length("this is a test\u10ffff") ; end
119 sec
```
Note: I'd like to have the last one come out as 119.31 sec, but I'd have to have a function build and return a string, instead of just an integer... not sure how best to do this
[warning: this code is NOT my best, just quick stuff to get it done, I need to move on to my "real" work,
but I still think it's as good as what was there, and hopefully should get moved into base :grinning: ]

**Comment by ScottPJones** [comment]
Please, bring on the virtual tomatoes or roses!
(I think I know how to handle getting the "floating point" output now, but I may save that for another PR...)

**Comment by ScottPJones** [comment]
What I *really* want for the output is the following:
For 1-999 nanoseconds, say `"%d nsec"` (should that be spelled out instead? nanoseconds?)
For 1000-999999 nanoseconds, say `"%d.%03d µsec"` (same thing, better microseconds?)
For 1000000-999999999 nanoseconds, say `"%d.%03d msec"` (milliseconds?) (rounding to the nearest microsecond)...
For 1 second to 999 seconds, `"%d.%03d sec"` (seconds?) (rounded to nearest millisecond)
Above 999 seconds, just put `"%d sec"`, rounded to nearest second.
Does that seem reasonable?

**Comment by IainNZ** [comment]
The use of `sec` seems non-standard: either `µs`, scientific notation-seconds, or fully spelled our `microseconds` would seem better. Also `11908 µsec` does my head in, if you are going to break it down by 1000s then thats `11.91 ms` (to 2.d.p, say)

**Comment by ScottPJones** [comment]
@IainNZ As I said in my last message, that `11908 µsec` wasn't how I really wanted to have it... I just hadn't yet figured out how to do what I wanted in Julia...  I actually did go ahead and change that already... Also, the short names for milliseconds etc. were because initially, the line could get too long, but it's now short enough that I think I can just spell everything out...  I'm outputting to 3 dp, makes it easier to compare between runs...

**Comment by IainNZ** [comment]
Maybe my package https://github.com/IainNZ/Humanize.jl/blob/master/src/Humanize.jl can provide inspiration?

**Comment by ScottPJones** [comment]
@IainNZ I'd asked back 3 days ago here if there were any pretty printing facilities like that, but nobody responded :cry:  I'll look at that... what do you think of my last go round?

**Comment by ScottPJones** [comment]
@IainNZ I looked at your Humanize, looks nice, I did go a bit differently with " bytes" instead of " B", I think while most people understand KB, MB, GB, TB, etc, "bytes" looked a lot better than just "B"...

**Comment by ScottPJones** [comment]
Anything else keeping this from being merged in to base?  Thanks!

**Comment by IainNZ** [comment]
I think it'd be helpful to squash the history, and post a sample of the output - the thread is very long and it has changed a lot, so its hard to review.

**Comment by kmsquire** [comment]
Actually, squashing the history of a PR with multiple commits is generally required before merging into base.

**Comment by ScottPJones** [comment]
OK, how do I squash the history? (Sorry, still only been a few weeks using git and GitHub)

**Comment by jakebolewski** [comment]
@ScottPJones look at the man page `git rebase --interactive`.

**Comment by timholy** [comment]
See also http://docs.julialang.org/en/latest/manual/packages/#squashing-and-rebasing

**Comment by ScottPJones** [comment]
Thanks!  As soon as I get back from heckling a poor PhD candidate, I'll take care of that! :grinning:

**Comment by ScottPJones** [comment]
Ok, I *tried* to squash this, and now I've got other people's commits... (I'm getting very frustrated with git & GitHub!)...

**Comment by ScottPJones** [comment]
Note: when I compare this branch to Julia, it only shows my commits...

**Comment by tkelman** [comment]
The idea is you want just 7582922a5c82de9a6ccd6e4bd51b4c34b02b24d8, yes? I suspect you had everything right until you tried to do a push, which complained and made you do a merge - the merge commit is where things get messy and you don't want to include those here.

Try the following
```
git checkout -b spj/time2 # save the current state to a new branch
git branch -D spj/time # delete existing messy copy
git checkout master
git checkout -b spj/time # make a new clean branch with the same name
git cherry-pick 7582922a5c82de9a6ccd6e4bd51b4c34b02b24d8 # add in the one commit you want
git push origin +spj/time # do a *force push* (marked by the + before the branch name) to overwrite the branch's contents with a hopefully clean replacement
```

**Comment by ScottPJones** [comment]
Thanks, is that correct now?  I've put some examples on Gist:
https://gist.github.com/ScottPJones/e7c38085a3c391578261
https://gist.github.com/ScottPJones/bb712f7b85d1d8d91a9a#file-results-from-v-0-4-time

**Comment by tkelman** [comment]
It's only one commit now which is good, you tell me if it has all the changes you wanted.

One superficial thing is it looks like you've got some mixed tabs and spaces, we try to avoid tabs everywhere except in makefiles, 4-space indent is preferred for the C and Julia code in src and base.

**Comment by ScottPJones** [comment]
Sorry about the tabs... I keep resetting the "smart-tabs" variable in the editor, and writing out the state, but it's like a zombie, it keeps coming back!  I did do 4-space indents everywhere (that's also what we came up with where I'm consulting)

**Comment by ScottPJones** [comment]
Do I need to squash this puppy again? :grinning:

**Comment by jakebolewski** [comment]
yes

**Comment by ScottPJones** [comment]
The puppy is all squished! :dog:

**Comment by tkelman** [comment]
Anyone else think `getunits` and `printnz` are somewhat likely to collide with future names that we may want to use elsewhere? Maybe not such a big deal since they're not exported from Base, but any future additions where those names would be natural to use will have to be aware of these methods.

And sorry to nitpick, but the couple times in the C code where you're using code like this
```
if (sz < old) freed_bytes += (old - sz);
else allocd_bytes += (sz - old);
```
could use some newlines and indenting to fit with the code style elsewhere.

**Comment by ScottPJones** [comment]
@tkelman As I've said before, I greatly appreciate all constructive criticism.  I'll change the formatting of the bug fixes with reallocs.  I just noticed the bugs and fixed them quickly, and 35 years of the style I've become accustomed to came out of my fingers without thought!   About the names, I'll change those
(do you have any suggestions?), but that actually brings up a major problem I have with Julia... it seems there is no way to keep your abstraction and implementation separate, and keep voyeurs from looking at your private parts, or indeed from playing with them... and everybody is also very exhibitionist... letting stuff that should stay private getting seen in public...  CLU was the best language I knew in that regard...

**Comment by tkelman** [comment]
> it seems there is no way to keep your abstraction and implementation separate [might want to edit the rest of this sentence to be slightly less colorful]

Right, there have been a few conversations around this, and other than sometimes sort-of following some of the Python convention of underscore-oriented-programming (which some people understandably dislike), I'm not sure we have a great solution for this yet. At least the convention is generally towards extension via methods and dispatch rather than monkey-patching the way Ruby and Python code often does.

Shipping hidden proprietary code written in Julia will probably have to be done by AOT compilation into shared libraries. Not sure whether there will be any better solutions but it's worth thinking about.

**Comment by ScottPJones** [comment]
I'm not concerned so much about shipping hidden proprietary code, but rather software engineering practices.  I've seen too many times what happens when customers start playing around directly with your implementation, and then you can be locked into that implementation forever (even if it sucks!).
Just like immutable keeps things from being modified, you need something to keep something private, so that you can't cause namespace pollution, or have to pick ugly long names to avoid it, which seems to be what is being suggested here for getunits and printnz.

**Comment by ScottPJones** [comment]
OK, please look at this again... I changed the names to `prettyprint_getunits` & `padded_nonzero_print`,
and changed the C code... (just saying, that style was prohibited in the C style guide I wrote! :grinning:
Our guide said that if the statement was not on the same line as the if/else/else if, then it needed { }.)

**Comment by ScottPJones** [comment]
OK... any hope of getting it merged? :grinning:  Thanks for all the comments!

**Comment by tkelman** [comment]
I think this is good enough to merge for now, we can keep tweaking it on master. Could do with some tests eventually, for one thing (but I don't think we have any for these macros now, so this isn't really making things any much worse).

**Comment by ScottPJones** [comment]
Thanks!  I’m not sure how you’d test a macro like this... @timed maybe, where it returns the information instead of printing it... if anyone has a suggestion, I’ll certainly add testing...

**Comment by yuyichao** [comment]
@ScottPJones  The time seems to be printed twice for `@timev`, is this intentional?

**Comment by ScottPJones** [comment]
Yes, the first line is the pretty printed value, exactly the same as @time, after that, it prints out the raw information that it collects, not rounded or anything...

**Comment by tkelman** [comment]
We do still need docs for the new `@timev` macro.

**Comment by ScottPJones** [comment]
Ah!  Yes, that slipped between the cracks... will do (must be that early onset Alzheimers!)
